### DISCUSSION TITLE
Using React

### DISCUSSION THREAD

**Discussion by nexensys** [ScratchAddons/ScratchAddons#3558] [Help]
So, I asked a while ago(when I knew nothing of react) if there would be any objections if I used Vue for my addon, and got some including the fact that there be many incompatibilities. Now I am asking If I would be okay to load `react` and `react-dom` from the prebuilt scripts. Since Scratch itself is bundled with webpack, this should not interfere, and I will not be mounting to an element that will be managed by Scratch's react. I am asking this because I am about ready to start building the GUI for my `text-editor` and It would GREATLY speed up the website, since I would either have to build my own Virtual Dom which, let's face it, would be nowhere near as efficient or I would have to completely re-render the entire script, which, for much larger projects, would be kind of dangerous, processing wise.

**Comment by WorldLanguages** [comment]
Scratch Addons isn't made with those types of addons in mind, at least for now.

If your addon is so complex it requires virtual DOM, we'd need to choose between these 3 options, and all are bad:
- Ship a virtual DOM library as part of the extension to all users, increasing the bundle size. And JSX wouldn't be supported.
- Add a build step to Scratch Addons as a whole because some addons need a build step.
- Allow some addons to ship preprocessed code to Scratch Addons (color picker library being the exception).

**Reply by nexensys** [reply]
> Ship a virtual DOM library as part of the extension to all users, increasing the bundle size. And JSX wouldn't be supported.

I wasn't expecting JSX, I just need `createElement`

> Add a build step to Scratch Addons as a whole because some addons need a build step.

I was actually already going to propose that, as it would greatly decrease load time and download size. Then some addons could use npm modules.

> Allow some addons to ship preprocessed code to Scratch Addons (color picker library being the exception).

What exactly do you mean?

**Reply by WorldLanguages** [reply]
> I was actually already going to propose that, as it would greatly decrease load time and download size. Then some addons could use npm modules.

But it would increase complexity when developing, and it's a compromise I don't think we should get into. Building an addon shouldn't be much harder than making a script with Greasemonkey.

> What exactly do you mean?

Well, addons that need a build step could reside in a different repo, and whenever needed, the main repo could be updated with the built version of the addon. But that would be problematic for various reasons.

**Reply by nexensys** [reply]
> But it would increase complexity when developing, and it's a compromise I don't think we should get into. Building an addon shouldn't be much harder than making a script with Greasemonkey.

Okay. So, no npm modules. But I still think that we should bundle before publishing, while still having the code work without bundling. But that would be super complex. Some sort of script that writes a new file for webpack based off of the `addons.json`

> Well, addons that need a build step could reside in a different repo, and whenever needed, the main repo could be updated with the built version of the addon. But that would be problematic for various reasons.

Ah. Ok. I can see the danger in that.

**Reply by nexensys** [reply]
Also, I forgot to say earlier, the reason that I want to use react so badly is that I am going to have a set of elements for every single script in the project. Now, this may not be so bad for a small project, but if one of griffpatch's projects was loaded, for example, there could be major issues.

**Reply by TheColaber** [reply]
Can't you use Blockly then? And what are you making... Why do you need an element for each script?

**Reply by nexensys** [reply]
I am making the visual portion of my text editor addon. I am going to have a new tab with a sort of pseudo-workspace in which there will be a text box with a title bar for each script so that they can be seen in the right places, and can be moved.

**Reply by nexensys** [reply]
And I am going to be using blockly, to read the workspace and create the scripts.

**Comment by nexensys** [comment]
Any decisions yet?

**Reply by WorldLanguages** [reply]
What would the `text-editor` addon actually do? Maybe we should discuss if we'd add an addon like that first?

**Reply by mxmou** [reply]
> What would the `text-editor` addon actually do? Maybe we should discuss if we'd add an addon like that first?

There was an issue about it - I think it's a way to edit scripts as text.

**Reply by nexensys** [reply]
> There was an issue about it - I think it's a way to edit scripts as text.

Yes. My parser is going to be a (moderately) adapted form of scratchblocks.

And I would like to have a bit of a visual interface to make it less ugly than a plain textarea. Hence React

**Reply by CubeyTheCube** [reply]
> > There was an issue about it - I think it's a way to edit scripts as text.
> 
> Yes. My parser is going to be a (moderately) adapted form of scratchblocks.
> 
> And I would like to have a bit of a visual interface to make it less ugly than a plain textarea. Hence React

Can't you use something like codemirror instead of reinventing the wheel?

**Reply by nexensys** [reply]
> Can't you use something like codemirror instead of reinventing the wheel?

Problem 1: Codemirror does not natively support it.

Problem 2: I can't just put all of the scripts in one large string of text. I am going to have multiple text boxes that are draggable and match the position of the blocks.

**Reply by WorldLanguages** [reply]
I'm not entirely sure if that should be an addon. You're essentially creating your own language, right?

**Reply by cobaltt7** [reply]
Yeah, no. Just use [tosh](https://tosh.blob.codes/) (crated by the same person who made scratchblocks)

**Reply by Dhuls-Scratch** [reply]
> Yeah, no. Just use [tosh](https://tosh.blob.codes/) (crated by the same person who made scratchblocks)

Unfortunately, tosh is still on 2.0, and it uses Phosphorus.

**Reply by nexensys** [reply]
> I'm not entirely sure if that should be an addon. You're essentially creating your own language, right?

No. It will have the same exact structure as scratchblocks just without overrides and the special scratchblocks blocks such as `. . .`

**Reply by nexensys** [reply]
*Almost* up to date version here: https://github.com/ErrorGamer2000/Scratch-Addons-Text-Editor

**Reply by nexensys** [reply]
Commits may be in a slightly confusing order, but I am trying to make them as sensible as possible

**Reply by WorldLanguages** [reply]
Hmm I don't really have the time to type a lot today, but my intuition dislikes the idea of a text editor as an addon. It would be the first addon that's not "straight forward" where we need to take an enormous amount of design decisions. It makes way more sense to fork Scratch and make it web based like tosh, first.

**Reply by nexensys** [reply]
Alright then...
